File: backend/app/schemas/resource_dashboard.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


DEFAULT_RESOURCE_DASHBOARD_NAMESPACE = "vllm-project"


def normalize_namespaces_value(value: str | None) -> str:
    namespaces = []
    seen = set()
    for item in (value or "").split(","):
        namespace = item.strip()
        if namespace and namespace not in seen:
            namespaces.append(namespace)
            seen.add(namespace)
    return ",".join(namespaces) if namespaces else DEFAULT_RESOURCE_DASHBOARD_NAMESPACE


class KubernetesClusterBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    context: str | None = Field(None, max_length=200)
    default_label_selector: str | None = Field(None, max_length=500)
    namespaces: str = Field("vllm-project", min_length=1, max_length=500, description="逗号分隔的命名空间列表")
    npu_resource_name: str = Field("huawei.com/Ascend910", min_length=1, max_length=200)
    enabled: bool = True
    display_order: int = 0

    @field_validator("default_label_selector", "context")
    @classmethod
    def normalize_empty_string(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None

    @field_validator("namespaces", mode="before")
    @classmethod
    def normalize_namespaces(cls, value: str) -> str:
        return normalize_namespaces_value(value)


class KubernetesClusterCreate(KubernetesClusterBase):
    kubeconfig: str = Field(..., min_length=1)


class KubernetesClusterUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    kubeconfig: str | None = None
    context: str | None = Field(None, max_length=200)
    default_label_selector: str | None = Field(None, max_length=500)
    namespaces: str | None = Field(None, min_length=1, max_length=500)
    npu_resource_name: str | None = Field(None, min_length=1, max_length=200)
    enabled: bool | None = None
    display_order: int | None = None

    @field_validator("kubeconfig", "default_label_selector", "context")
    @classmethod
    def normalize_empty_string(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None

    @field_validator("namespaces", mode="before")
    @classmethod
    def normalize_namespaces(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return normalize_namespaces_value(value)


class KubernetesClusterResponse(KubernetesClusterBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    kubeconfig_configured: bool = True
    created_by: int | None = None
    created_at: datetime
    updated_at: datetime


class KubernetesClusterTestResponse(BaseModel):
    success: bool
    message: str
    node_count: int = 0
    pod_count: int = 0


class ResourceQuantity(BaseModel):
    cpu_cores: float = 0
    memory_bytes: float = 0
    npu: float = 0


class ResourceNodeInfo(BaseModel):
    node_name: str
    total: ResourceQuantity
    used: ResourceQuantity
    available: ResourceQuantity
    running_instances: int = 0
    executing_pods_count: int = 0


class ResourcePodInfo(BaseModel):
    cluster_id: int
    cluster_name: str
    namespace: str
    name: str
    phase: str | None = None
    node_name: str | None = None
    labels: dict[str, str] = Field(default_factory=dict)
    created_at: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_seconds: int | None = None
    pr_number: int | None = None
    pr_url: str | None = None
    job_workflow_ref: str | None = None
    requests: ResourceQuantity
    containers: list[str] = Field(default_factory=list)


class ClusterResourceSummary(BaseModel):
    cluster_id: int
    cluster_name: str
    total: ResourceQuantity
    used: ResourceQuantity
    available: ResourceQuantity
    running_instances: int = 0
    executing_pods_count: int = 0
    executed_pods_count: int = 0
    node_resources: list[ResourceNodeInfo] = Field(default_factory=list)
    executing_pods: list[ResourcePodInfo] = Field(default_factory=list)
    scope: dict[str, Any]
    error: str | None = None


class ResourceDashboardResponse(BaseModel):
    overall: ClusterResourceSummary
    clusters: list[ClusterResourceSummary]
    executing_pods: list[ResourcePodInfo]
    executed_pods: list[ResourcePodInfo]

File: frontend/src/services/resourceDashboard.ts
import api from './api'

export interface KubernetesCluster {
  id: number
  name: string
  description?: string | null
  context?: string | null
  default_label_selector?: string | null
  npu_resource_name: string
  enabled: boolean
  display_order: number
  kubeconfig_configured: boolean
  created_by?: number | null
  created_at: string
  updated_at: string
}

export interface KubernetesClusterCreate {
  name: string
  description?: string | null
  kubeconfig: string
  context?: string | null
  default_label_selector?: string | null
  npu_resource_name: string
  enabled: boolean
  display_order: number
}

export interface KubernetesClusterUpdate {
  name?: string
  description?: string | null
  kubeconfig?: string | null
  context?: string | null
  default_label_selector?: string | null
  npu_resource_name?: string
  enabled?: boolean
  display_order?: number
}

export interface KubernetesClusterTestResponse {
  success: boolean
  message: string
  node_count: number
  pod_count: number
}

export interface ResourceQuantity {
  cpu_cores: number
  memory_bytes: number
  npu: number
}

export interface ResourceNodeInfo {
  node_name: string
  total: ResourceQuantity
  used: ResourceQuantity
  available: ResourceQuantity
  running_instances: number
  executing_pods_count: number
}

export interface ClusterResourceSummary {
  cluster_id: number
  cluster_name: string
  total: ResourceQuantity
  used: ResourceQuantity
  available: ResourceQuantity
  running_instances: number
  executing_pods_count: number
  executed_pods_count: number
  node_resources: ResourceNodeInfo[]
  scope: Record<string, unknown>
  error?: string | null
}

export interface ResourcePodInfo {
  cluster_id: number
  cluster_name: string
  namespace: string
  name: string
  phase?: string | null
  status?: string | null
  node_name?: string | null
  labels: Record<string, string>
  created_at?: string | null
  started_at?: string | null
  finished_at?: string | null
  duration_seconds?: number | null
  pr_number?: number | null
  pr_url?: string | null
  job_workflow_ref?: string | null
  requests: ResourceQuantity
  containers: string[]
}

export interface ResourceDashboardResponse {
  generated_at: string
  overall: ClusterResourceSummary
  clusters: ClusterResourceSummary[]
  executing_pods: ResourcePodInfo[]
  executed_pods: ResourcePodInfo[]
}

export interface ResourceDashboardParams {
  cluster_ids?: number[]
  label_selector?: string
  include_pods?: boolean
}

const buildParams = (params: ResourceDashboardParams) => {
  const searchParams = new URLSearchParams()
  params.cluster_ids?.forEach(id => searchParams.append('cluster_ids', String(id)))
  if (params.label_selector) {
    searchParams.append('label_selector', params.label_selector)
  }
  if (params.include_pods !== undefined) {
    searchParams.append('include_pods', String(params.include_pods))
  }
  return searchParams
}

export const getEnabledResourceClusters = async () => {
  const response = await api.get<KubernetesCluster[]>('/resource-dashboard/clusters/enabled')
  return response.data
}

export const getResourceDashboard = async (params: ResourceDashboardParams = {}) => {
  const response = await api.get<ResourceDashboardResponse>('/resource-dashboard/summary', {
    params: buildParams(params),
  })
  return response.data
}

export const listResourceClusters = async () => {
  const response = await api.get<KubernetesCluster[]>('/resource-dashboard/clusters')
  return response.data
}

export const createResourceCluster = async (data: KubernetesClusterCreate) => {
  const response = await api.post<KubernetesCluster>('/resource-dashboard/clusters', data)
  return response.data
}

export const updateResourceCluster = async (id: number, data: KubernetesClusterUpdate) => {
  const response = await api.put<KubernetesCluster>(`/resource-dashboard/clusters/${id}`, data)
  return response.data
}

export const deleteResourceCluster = async (id: number) => {
  const response = await api.delete<{ message: string }>(`/resource-dashboard/clusters/${id}`)
  return response.data
}

export const testResourceCluster = async (id: number) => {
  const response = await api.post<KubernetesClusterTestResponse>(`/resource-dashboard/clusters/${id}/test`)
  return response.data
}
